File: samples/java/TicTacToe/Server/src/main/java/systems/zlink/samples/tictactoe/server/configuration/TicTacToeReadinessReporter.java
package systems.zlink.samples.tictactoe.server.configuration;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.ApplicationArguments;
import org.springframework.boot.ApplicationRunner;

import systems.zlink.contracts.core.RoutingId;
import systems.zlink.framework.errors.ZLinkConfigurationException;
import systems.zlink.framework.monitoring.ZLinkPeerState;
import systems.zlink.framework.monitoring.ZLinkRouteMeshRuntime;

import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

public final class TicTacToeReadinessReporter implements ApplicationRunner, AutoCloseable {
    private static final Logger LOGGER = LoggerFactory.getLogger(TicTacToeReadinessReporter.class);
    private final String nodeId;
    private final String peerNodeId;
    private final RoutingId expectedPeerRoutingId;
    private final ZLinkRouteMeshRuntime meshes;
    private final ScheduledExecutorService reporter =
            Executors.newSingleThreadScheduledExecutor(
                    runnable -> {
                        Thread thread = new Thread(runnable, "tictactoe-readiness");
                        thread.setDaemon(true);
                        return thread;
                    });

    private TicTacToeReadinessReporter(
            String nodeId,
            String peerNodeId,
            RoutingId expectedPeerRoutingId,
            ZLinkRouteMeshRuntime meshes) {
        this.nodeId = nodeId;
        this.peerNodeId = peerNodeId;
        this.expectedPeerRoutingId = expectedPeerRoutingId;
        this.meshes = meshes;
    }

    public static TicTacToeReadinessReporter peerRoute(
            String nodeId, String peerNodeId, ZLinkRouteMeshRuntime meshes) {
        return new TicTacToeReadinessReporter(
                nodeId, peerNodeId, RoutingId.from("tictactoe-" + peerNodeId), meshes);
    }

    public static TicTacToeReadinessReporter spotRoute(
            String nodeId, ZLinkRouteMeshRuntime meshes) {
        return new TicTacToeReadinessReporter(nodeId, null, null, meshes);
    }

    @Override
    public void run(ApplicationArguments arguments) {
        reporter.scheduleWithFixedDelay(this::report, 0, 100, TimeUnit.MILLISECONDS);
    }

    private void report() {
        try {
            var snapshot = meshes.snapshot(SampleNames.SpotMesh);
            boolean ready =
                    expectedPeerRoutingId == null
                            ? snapshot.readyPeerCount() >= 2
                            : snapshot.peers().stream()
                                    .anyMatch(
                                            peer ->
                                                    peer.nodeRid().equals(expectedPeerRoutingId)
                                                            && peer.state()
                                                                    == ZLinkPeerState.READY);
            if (!ready) return;
            if (expectedPeerRoutingId == null) {
                LOGGER.info(
                        "tictactoe-ready kind=spot-route node={} mesh={}",
                        nodeId,
                        SampleNames.SpotMesh);
            } else {
                LOGGER.info("tictactoe-ready kind=peer-route node={} peer={}", nodeId, peerNodeId);
            }
            reporter.shutdown();
        } catch (IllegalStateException | ZLinkConfigurationException ignored) {
            // The public runtime view is not available until Framework startup completes.
        }
    }

    @Override
    public void close() {
        reporter.shutdownNow();
    }
}

File: samples/kotlin/TicTacToe/Server/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/server/configuration/TicTacToeReadinessReporter.kt
package systems.zlink.samples.kotlin.tictactoe.server.configuration

import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import org.slf4j.LoggerFactory
import org.springframework.boot.ApplicationArguments
import org.springframework.boot.ApplicationRunner
import systems.zlink.framework.errors.ZLinkConfigurationException
import systems.zlink.framework.monitoring.ZLinkPeerState
import systems.zlink.framework.monitoring.ZLinkRouteMeshRuntime

class TicTacToeReadinessReporter private constructor(private val checks: List<ReadinessCheck>) :
    ApplicationRunner, AutoCloseable {
    private val reporter =
        Executors.newSingleThreadScheduledExecutor { runnable ->
            Thread(runnable, "tictactoe-readiness").apply { isDaemon = true }
        }

    override fun run(args: ApplicationArguments) {
        reporter.scheduleWithFixedDelay(::report, 0, 100, TimeUnit.MILLISECONDS)
    }

    private fun report() {
        checks.forEach(ReadinessCheck::reportIfReady)
        if (checks.all(ReadinessCheck::reported)) {
            reporter.shutdown()
        }
    }

    override fun close() {
        reporter.shutdownNow()
    }

    private class ReadinessCheck(private val evidence: String, private val ready: () -> Boolean) {
        var reported: Boolean = false
            private set

        fun reportIfReady() {
            if (reported) return
            try {
                if (ready()) {
                    logger.info(evidence)
                    reported = true
                }
            } catch (_: IllegalStateException) {
                // The public runtime view is not available until Framework startup completes.
            } catch (_: ZLinkConfigurationException) {
                // The public runtime view is not available until Framework startup completes.
            }
        }
    }

    companion object {
        private val logger = LoggerFactory.getLogger(TicTacToeReadinessReporter::class.java)

        fun api(nodeId: String, meshes: ZLinkRouteMeshRuntime): TicTacToeReadinessReporter =
            TicTacToeReadinessReporter(
                listOf(
                    ReadinessCheck(
                        "tictactoe-ready kind=spot-route node=$nodeId mesh=${SampleNames.SpotMesh}"
                    ) {
                        meshes.snapshot(SampleNames.SpotMesh).isReady
                    }
                )
            )

        fun play(
            nodeId: String,
            peerNodeId: String,
            meshes: ZLinkRouteMeshRuntime,
        ): TicTacToeReadinessReporter {
            val peerRoutingId = "tictactoe-play-$peerNodeId"
            return TicTacToeReadinessReporter(
                listOf(
                    ReadinessCheck(
                        "tictactoe-ready kind=peer-route node=$nodeId peer=$peerNodeId"
                    ) {
                        meshes.snapshot(SampleNames.SpotMesh).peers().any { peer ->
                            peer.nodeRid().toString() == peerRoutingId &&
                                peer.state() == ZLinkPeerState.READY
                        }
                    }
                )
            )
        }
    }
}
